### PR TITLE
Update ecma_op_typedarray_from to use iterable objects

### PR DESCRIPTION
**Pull request by szilagyiadam**
Added steps 6-8 from ECMA-262 v6, 22.2.2.1.1

JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 1]
```diff
@@ -700,12 +701,97 @@ ecma_op_typedarray_from (ecma_value_t items_val, /**< the source array-like obje
...
+    /* 8.h */
+    if (ECMA_IS_VALUE_ERROR (new_typedarray))
+    {
+      ecma_collection_free (values_p);
+      return new_typedarray;
+    }
+
+    len = values_p->item_count;
+
+    goto create_typedarray;
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-typedarray-object.c]: Do we really need this goto? Can't we just refactor the code a bit? Maybe move some code into methods?

File: tests/jerry/es2015/typedarray-from.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                } else {
+                    this.index = 0;
+                    return {done: true}
+                }
+            }
+        }
+    }
+}
+
+var c = new SimpleClass([1,2,3,4,5]);
```
Comment (galpeter) [inline tests/jerry/es2015/typedarray-from.js]: Maybe I'm missing something, but how is this testcase related to the typeaddary.from?

File: jerry-core/ecma/operations/ecma-typedarray-object.c (hunk 1 above)
Comment (szilagyiadam) [reply jerry-core/ecma/operations/ecma-typedarray-object.c]: I used goto, because in the ECMA standard steps 8.j. and 17. are almost the same, the only difference is when we create the new typedarray from iterable object, we use collections. I can remove the goto and implement the steps from the standard, but that will increase binary size. What is your advice?

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 3]
```diff
@@ -700,12 +759,120 @@ ecma_op_typedarray_from (ecma_value_t items_val, /**< the source array-like obje
...
+                                                               values_p->buffer_p[index],
+                                                               index,
+                                                               func_object_p,
+                                                               info,
+                                                               setter_cb);
+
+      if (ECMA_IS_VALUE_ERROR (error_val))
+      {
+        ecma_deref_object (new_typedarray_p);
+        ecma_collection_destroy (values_p);
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-typedarray-object.c]: before the collection_destroy the remaining elements in the collection must be freed.

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 4]
```diff
@@ -700,12 +759,126 @@ ecma_op_typedarray_from (ecma_value_t items_val, /**< the source array-like obje
...
+
+        ecma_collection_destroy (values_p);
+        return error_val;
+      }
+    }
+
+    ecma_ref_object (new_typedarray_p);
+    ret_value = ecma_make_object_value (new_typedarray_p);
+
+    ecma_deref_object (new_typedarray_p);
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-typedarray-object.c]: IMO there is an unnecessary ref/deref call for new_typedarray_p.

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 5]
```diff
@@ -743,68 +915,41 @@ ecma_op_typedarray_from (ecma_value_t items_val, /**< the source array-like obje
...
-      ecma_value_t mapped_value;
-      if (func_object_p != NULL)
+      ecma_value_t error_val = ecma_op_typedarray_from_helper (this_val,
+                                                               current_value,
+                                                               index,
+                                                               func_object_p,
+                                                               info,
+                                                               setter_cb);
+
+      if (ECMA_IS_VALUE_ERROR (error_val))
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-typedarray-object.c]: This can be ORed (`||`) tho the condition above, and you can always return `ECMA_VALUE_ERROR`.

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 6]
```diff
@@ -682,7 +683,65 @@ ecma_typedarray_create_object_with_typedarray (ecma_object_t *typedarray_p, /**<
...
+  ecma_free_value (mapped_value);
+
+  if (ECMA_IS_VALUE_ERROR (mapped_number))
+  {
+    return mapped_number;
+  }
+
+  if (index >= info.length)
+  {
+    ecma_raise_type_error (ECMA_ERR_MSG ("Invalid argument type."));
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-typedarray-object.c]: you can return with `ecma_raise_type_error` directly.

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 7]
```diff
@@ -682,7 +683,65 @@ ecma_typedarray_create_object_with_typedarray (ecma_object_t *typedarray_p, /**<
...
+    return ECMA_VALUE_ERROR;
+  }
+
+  ecma_length_t byte_pos = (index << info.shift);
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-typedarray-object.c]: this local variable is not needed.

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 8]
```diff
@@ -700,12 +759,120 @@ ecma_op_typedarray_from (ecma_value_t items_val, /**< the source array-like obje
...
+    ecma_typedarray_setter_fn_t setter_cb = ecma_get_typedarray_setter_fn (info.id);
+
+    /* 8.j */
+    for (uint32_t index = 0; index < values_p->item_count; index++)
+    {
+      ecma_value_t error_val = ecma_op_typedarray_from_helper (this_val,
+                                                               values_p->buffer_p[index],
+                                                               index,
+                                                               func_object_p,
+                                                               info,
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-typedarray-object.c]: I'd pass this structure as reference.

File: jerry-core/ecma/operations/ecma-typedarray-object.c [hunk 9]
```diff
@@ -682,7 +683,64 @@ ecma_typedarray_create_object_with_typedarray (ecma_object_t *typedarray_p, /**<
...
+static ecma_value_t
+ecma_op_typedarray_from_helper (ecma_value_t this_val, /**< this_arg for the above from function */
+                                ecma_value_t current_value, /**< given value to set */
+                                uint32_t index, /**< currrent index */
+                                ecma_object_t *func_object_p, /**< map function object */
+                                ecma_typedarray_info_t *info_p, /**< typedarray info */
+                                ecma_typedarray_setter_fn_t setter_cb) /**< setter callback function */
+{
+  ecma_value_t mapped_value;
+  if (func_object_p != NULL)
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-typedarray-object.c]: AFAIK, this case is not tested. There should be a test which uses the "mapfn" from the standard.
